packages/ckeditor5-table/src/tableutils.js

```diff
@@ -132,37 +137,52 @@ export default class TableUtils extends Plugin {
 			}
 
 			// Inserting at the end and at the beginning of a table doesn't require to calculate anything special.
-			if ( insertAt === 0 || insertAt === table.childCount ) {
-				createEmptyRows( writer, table, insertAt, rowsToInsert, this.getColumns( table ) );
+			if ( copyStructureFrom < 0 && ( insertAt === 0 || insertAt === rows ) ) {
+				createEmptyRows( writer, table, insertAt, rowsToInsert, columns );
 
 				return;
 			}
 
-			// Iterate over all rows above inserted rows in order to check for rowspanned cells.
-			const tableIterator = new TableWalker( table, { endRow: insertAt } );
+			// Iterate over all rows above inserted rows in order to check for row-spanned cells.
+			const tableIterator = new TableWalker( table, { endRow: Math.max( insertAt, copyStructureFrom ) } );
 
 			// Will hold number of cells needed to insert in created rows.
-			// The number might be different then table cell width when there are rowspanned cells.
-			let cellsToInsert = 0;
+			// The number might be different then table cell width when there are row-spanned cells.
+			let cellsToInsert = columns;
+
+			// Store spans of the reference row to reproduce it's structure. This array is column number indexed.
+			const columnsSpans = new Array( columns ).fill( 1 );
 
-			for ( const { row, rowspan, colspan, cell } of tableIterator ) {
-				const isBeforeInsertedRow = row < insertAt;
-				const overlapsInsertedRow = row + rowspan > insertAt;
+			for ( const { row, column, rowspan, colspan, cell } of tableIterator ) {
+				const lastCellRow = row + rowspan - 1;
 
-				if ( isBeforeInsertedRow && overlapsInsertedRow ) {
+				if ( row < insertAt && insertAt <= lastCellRow ) {
```

Comment:
AFAICS the logic here wasn't changed. Please do not remove helper `is...` from `if` - they helps to read what that case handles. We may know at this point in time why there is `<` and what we compare here but after a while we might forget.